Improve Cashbox Report

jniles opened the pull request:
This PR improves the cashbox report in the following ways:
 1. The dates (from - to) are now respected in the search query.
 2. The currencies are rendered in the correct cashbox currency.
 3. We include a real opening balance and closing balance.
 4. The user can identify the difference between posted and unposted records via italics applied to unposted records.

We've also centralized the `getAccountTransactions()` call used between multiple reports into a single shared file `accounts/transactions`.  We've also rewritten the query for the combined posting_journal and general_ledger such that the WHERE statement is executed in subqueries, greatly improving the performance of the combined query.

Finally, the `getAccountTransactions()` function now sorts first based on `created_at` date, then on `trans_date`.  This makes back-dated transactions show up later in the rendering.  For example, users frequently back-date the transfer from the cashbox into the main coffre, and it makes sense to show these at the bottom of the report.

Closes #3595.

jniles committed:
refactor(accounts): mv getTransactions() into file

We create a separate file for the getAccountTransactions() function so
that it may be more easily imported and shared between reports.  The
account report and multiple accounts report have been updated to reflect
this new location.

jniles committed:
feat(reports): cash report uses accounts/txns.js

This commit migrates the cash report to using the account/transactions
controller to look up values.  The performance of accounts/transactions
has also been greatly increased by moving the WHERE statement into the
subqueries and generating subqueries.

jniles committed:
fix(report): add parens to account subquery

This commit adds parentheses to the subquery in the account report.

jniles committed:
refactor(reports): rework separated cash report

This commit reworks the separated income/expense portions of the cash
report so that they are more streamlined.  The code is easier to read
both in the template and the controller.  Further, the ids are changed
from numbers to text strings which greatly improves readability.

jniles committed:
feat(report): add balance to cash report

This commit adds the balance to the cash report when both the income and
expenses are visible on the balance report.

jniles committed:
feat(report): add transaction type to cash report

This commit adds the transaction type to the cash report.

mbayopanda commented on server/controllers/finance/accounts/transactions.js:
```diff
@@ -0,0 +1,225 @@
+const db = require('../../../lib/db');
+const FilterParser = require('../../../lib/filter');
+const Exchange = require('../exchange');
+const Accounts = require('.');
```
What does it means ? The current file or all files of the current directory.

If it is the current file, why do we need to import it ?

jniles replied:
Aha.  That is confusing.  It actually imports `./index.js`.  This file is `./transactions.js`.

mbayopanda commented on server/controllers/finance/accounts/transactions.js:
```diff
@@ -0,0 +1,225 @@
+  const enterpriseCurrencyColumns = `
+    (debit / rate) AS exchangedDebit, (credit / rate) AS exchangedCredit,
+    (balance / rate) AS exchangedBalance, @cumsum := (balance / rate) + @cumsum AS cumsum
+  `;
```
The request is so clear 👍

mbayopanda reviewed (commented):
LGTM 👍 

I was not able to test it with real data of imck because the cashbox selection did not work

jniles committed:
fix(report): allow any cashbox to be viewed

Removes the restriction on the cash report which loads only cashboxes
assigned to the user.
